File: frontend/types/pokemon.ts
export interface PokemonStats {
  hp: number;
  attack: number;
  defense: number;
  speed: number;
}

export interface PokemonData {
  id: number;
  name: string;
  types: string[];
  stats: PokemonStats;
  sprite: string;
  rarity: 'common' | 'uncommon' | 'rare' | 'legendary';
}

export interface PokemonNFT {
  id: string;
  speciesId: number;
  name: string;
  level: number;
  experience: number;
  stats: PokemonStats;
  types: string[];
  owner: string;
  mintTimestamp: number;
}

export interface EggNFT {
  id: string;
  parent1Species: number;
  parent2Species: number;
  incubationSteps: number;
  requiredSteps: number;
  genetics: number[];
  owner: string;
  createdTimestamp: number;
}

export interface BattleEvent {
  turn: number;
  attacker: string;
  defender: string;
  move: string;
  damage: number;
  effectiveness: number;
  commentary: string;
}

export interface Quest {
  id: string;
  title: string;
  description: string;
  objectives: QuestObjective[];
  rewards: QuestReward;
  expiresAt: Date;
}

export interface QuestObjective {
  type: 'battle' | 'capture' | 'hatch' | 'trade';
  target: number;
  current: number;
  description: string;
}

export interface QuestReward {
  type: 'tokens' | 'pokemon' | 'egg';
  amount?: number;
  pokemonId?: number;
}

export interface MarketplaceListing {
  id: string;
  listingId: string;
  nftId: string;
  nftType: 'pokemon' | 'egg';
  sellerAddress: string;
  price: number;
  nftMetadata: {
    species?: string;
    level?: number;
    stats?: PokemonStats;
    incubationProgress?: number;
    parentSpecies?: string[];
  };
  listedAt: Date;
  status: 'active' | 'sold' | 'cancelled';
  soldAt?: Date;
  buyerAddress?: string;
}

File: backend/models/pokemon.py
from pydantic import BaseModel, Field
from typing import List, Optional
from enum import Enum


class Rarity(str, Enum):
    COMMON = "common"
    UNCOMMON = "uncommon"
    RARE = "rare"
    LEGENDARY = "legendary"


class PokemonStats(BaseModel):
    hp: int = Field(..., ge=1, le=255)
    attack: int = Field(..., ge=1, le=255)
    defense: int = Field(..., ge=1, le=255)
    speed: int = Field(..., ge=1, le=255)


class PokemonData(BaseModel):
    id: int = Field(..., ge=1)
    name: str
    types: List[str]
    stats: PokemonStats
    sprite: str  # Front sprite (animated if available)
    back_sprite: Optional[str] = None  # Back sprite for battles
    rarity: Rarity


class PokemonNFT(BaseModel):
    id: str
    species_id: int
    name: str
    level: int = Field(..., ge=1, le=100)
    experience: int = Field(..., ge=0)
    stats: PokemonStats
    types: List[str]
    owner: str
    mint_timestamp: int


class EggNFT(BaseModel):
    id: str
    parent1_species: int
    parent2_species: int
    incubation_steps: int = Field(..., ge=0)
    required_steps: int = Field(default=1000)
    genetics: List[int]
    owner: str
    created_timestamp: int


class CaptureAttempt(BaseModel):
    pokemon_id: int
    health_percent: float = Field(..., ge=0.0, le=1.0)
    rarity: Rarity


class CaptureResult(BaseModel):
    success: bool
    capture_rate: float
    message: str
